lib/functions.sh

```diff
@@ -310,79 +315,101 @@ function get_or_create_role {
     echo $role_id
 }
 
-# Usage: get_or_add_user_project_role <role> <user> <project>
+# Returns the domain parts of a function call if present
+# Usage: _get_domain_args [<user_domain> <project_domain>]
+function _get_domain_args {
+    local domain
+    domain=""
+
+    if [[ -n "$1" ]]; then
+        domain="$domain --user-domain $1"
+    fi
+    if [[ -n "$2" ]]; then
+        domain="$domain --project-domain $2"
+    fi
+
+    echo $domain
+}
+
+# Gets or adds user role to project
+# Usage: get_or_add_user_project_role <role> <user> <project> [<user_domain> <project_domain>]
 function get_or_add_user_project_role {
     local user_role_id
+
+    domain_args=$(_get_domain_args $4 $5)
+
     # Gets user role id
     user_role_id=$(openstack role assignment list \
+        --role $1 \
         --user $2 \
         --project $3 \
-        --role $1 \
-        --column "Role" \
-        --format "value")
+        $domain_args \
+        | grep '^|\s[a-f0-9]\+' | get_field 1)
     if [[ -z "$user_role_id" ]]; then
         # Adds role to user and get it
         openstack role add $1 \
             --user $2 \
-            --project $3
+            --project $3 \
+            $domain_args
         user_role_id=$(openstack role assignment list \
+            --role $1 \
             --user $2 \
             --project $3 \
-            --role $1 \
-            --column "Role" \
-            --format "value")
+            $domain_args \
+            | grep '^|\s[a-f0-9]\+' | get_field 1)
     fi
     echo $user_role_id
 }
 
+# Gets or adds user role to domain
 # Usage: get_or_add_user_domain_role <role> <user> <domain>
 function get_or_add_user_domain_role {
     local user_role_id
     # Gets user role id
     user_role_id=$(openstack role assignment list \
+        --role $1 \
         --user $2 \
         --domain $3 \
-        --role $1 \
-        --column "Role" \
-        --format "value")
+        | grep '^|\s[a-f0-9]\+' | get_field 1)
```

Comment:
Why do this instead of using the --column and --format options to get the exact answer?

Reply:
@cmurphy because I just synced the functions from devstack (and I would prefer to keep it that way). Maybe fix it in devstack first and then we can sync again?